pv3_rust/src/pv.rs

```diff
@@ -0,0 +1,395 @@
+                buffer: std::ptr::null_mut(),
+                chunk_size: chunk_size_,
+                chunk_pool: Vec::with_capacity(chunk_count_ as usize),
+                chunk_pool_idx: chunk_count_,
+                fq: std::ptr::read(fq_ptr.cast::<xsk_ring_prod>()),
+                rx: std::ptr::read(rx_ptr.cast::<xsk_ring_cons>()),
+                cq: std::ptr::read(cq_ptr.cast::<xsk_ring_cons>()),
+                tx: std::ptr::read(tx_ptr.cast::<xsk_ring_prod>()),
+                }
+            };
+
+        a
+    }
+}
+
+fn _pv_alloc(nic: &mut PvNic) -> u64 {
+    if nic.chunk_pool_idx > 0 {
+        nic.chunk_pool_idx -= 1;
+        nic.chunk_pool[nic.chunk_pool_idx as usize]
+    } else {
+        return u64::MAX;
+    }
+}
+
+fn _pv_free(nic: &mut PvNic, chunk_addr: u64) {
+    // align **chunk_addr with chunk size
+    let remainder = chunk_addr % (nic.chunk_size as u64);
+    let chunk_addr = chunk_addr - remainder;
+    nic.chunk_pool[nic.chunk_pool_idx as usize] = chunk_addr;
+    nic.chunk_pool_idx += 1;
+}
+
+/* TODO: packet_process() 구현하면서 remove를 사용해서 벡터중 해당 요소만 제거하는 것으로 소스짜기 (WIP)*/
+pub fn pv_free(nic: &mut PvNic, packets: &mut Vec<PvPacket>) {
+    let packet_count: usize = packets.len();
+
+    for i in 0..packet_count {
+        _pv_free(nic, packets[packet_count - 1 - i].private as u64);
+    }
+    packets.clear();
+    // packets.remove(1); // WIP
+
+}
+
+fn configure_umem(nic: &mut PvNic, chunk_size: u32, chunk_count: u32, fill_size: u32, complete_size: u32) -> i32 {
+    let umem_buffer_size: usize = (chunk_count * chunk_size) as usize;
+    let mmap_address = unsafe {
+        libc::mmap(std::ptr::null_mut(), umem_buffer_size,
+                    libc::PROT_READ | libc:: PROT_WRITE,
+                libc::MAP_PRIVATE | libc::MAP_ANONYMOUS, -1, 0)
+    };
+
+    let umem_cfg: xsk_umem_config = xsk_umem_config {
+        fill_size: fill_size,
+        comp_size: complete_size,
+        frame_size: chunk_size,
+        frame_headroom: XSK_UMEM__DEFAULT_FRAME_HEADROOM,
+        flags: XSK_UMEM__DEFAULT_FLAGS,
+    };
+
+    let ret:c_int;
+
+    unsafe {
+        ret = xsk_umem__create(&mut nic.umem, mmap_address, umem_buffer_size as u64, &mut nic.fq, &mut nic.cq, &umem_cfg);
+    }
+
+    nic.buffer = mmap_address;
+    ret
+}
+
+pub fn pv_open(if_name: String, chunk_size: u32, chunk_count: u32,
+                rx_ring_size: u32, tx_ring_size: u32, filling_ring_size: u32,
+                completion_ring_size: u32) -> Option<PvNic>
+{
+    /* create PvNic */
+    let mut nic: PvNic = PvNic::new(chunk_size, chunk_count as i32);
+
+    /* initialize UMEM chunk information */
+    for i in 0..chunk_count {
+        nic.chunk_pool.push((i * chunk_size) as u64);   // put chunk address
+    }
+
+    /* configure UMEM */
+    let ret = configure_umem(&mut nic, chunk_size, chunk_count, filling_ring_size, completion_ring_size);
```

Comment:
If `configure_umem` returns `Result`, this could be prettier that matches rust style

Reply:
Nice idea. The suggestion is reflected